test/e2e/run_test.go

```diff
@@ -48,6 +48,26 @@ var _ = Describe("Podman run", func() {
 		Expect(session).Should(Exit(0))
 	})
 
+	It("podman run two containers choose different names", func() {
```

Comment:
Misleading: should be **container IDs**, not names. The output of `run -d` is a CID.

Unless you really want to test container **names**, in which case you'll need to `podman ps` or `podman inspect` or something else. 

But I'm not sure there's any point to this test: it seems impossible for two containers to have the same CID or name.

Reply:
https://github.com/containers/podman/blob/0d171e9b7a9857e4809037c152ed1c617e5b5a79/libpod/container_internal.go#L510-L514 suggests that this is names.

---

> It seems impossible for two containers to have the same CID or name.

So I think a test could create a container, record the name, delete it; create another container, record the name, delete it. Then check that the two generated names are different, i.e. that we don’t always generate the same (sequence of) names.

That would actually, unlike the previous unit test ( https://github.com/containers/podman/pull/18326#discussion_r1176766233 ), be a representative test of the whole-process RNG seeding behavior.

OTOH, it would also be an inherently flaky test, possibly generating the same name in the two runs purely by chance of how the RNG is seeded.

Reply:
>suggests that this is names

I'm very confused. I see the test running `podman run -d`, which AFAIK only emits CIDs and has only ever emitted CIDs, as has docker. I don't recall ever seeing a name.

But anyhow.  I can't get podman to reuse names even if I hardcode the seed:
```diff
diff --git a/libpod/runtime_pre_go1.20.go b/libpod/runtime_pre_go1.20.go
index 1c73043de..650e5e8d0 100644
--- a/libpod/runtime_pre_go1.20.go
+++ b/libpod/runtime_pre_go1.20.go
@@ -7,12 +7,11 @@ package libpod
 
 import (
 	"math/rand"
-	"time"
 )
 
 func init() {
 	// generateName calls namesgenerator.GetRandomName which the
 	// global RNG from math/rand. Seed it here to make sure we
 	// don't get the same name every time.
-	rand.Seed(time.Now().UnixNano())
+	rand.Seed(1024)
 }

```
```console
$ go version
go version go1.19.8 linux/amd64

$ for i in $(seq 1 10);do bin/podman run -d quay.io/libpod/testimage:20221018 top;done
...
$ bin/podman ps --format '{{.Names}}'|sort
blissful_feynman
cool_albattani
determined_colden
gifted_bhabha
gifted_thompson
magical_goldwasser
nifty_margulis
romantic_moore
sad_cartwright
wizardly_shaw
```

Reply:
I’m sorry, I just meant that the thing we _want_ to test is generated names.

You’re right that this currently tests IDs.

---

> I can't get podman to reuse names even if I hardcode the seed:

Looking at the current `main`, there is also https://github.com/containers/podman/blob/0277adf69363da9bf6f92c5eb5c2e7d99a5e2446/vendor/github.com/containerd/stargz-snapshotter/estargz/testutil.go#L48-L50 (which, despite the `test` in the name, does, I think run in Podman production). That might be the cause, but I didn’t test in practice or debug further.

Reply:
Separately,

> `for i in $(seq 1 10);do bin/podman run -d quay.io/libpod/testimage:20221018 top`

is it possible that older containers are still running while the loop proceeds to another iteration? (I didn’t test to confirm I’m afraid.)

In that case Podman would continue to try generating new names until there is no conflict (IOW, the non-deterministic seeding should improve performance to avoid retries, and make other corner cases rarer; it’s not because a conflict should immediately cause a failure).

Reply:
Sorry for the confusion, the test I wrote is relying on the fact that podman will barf when trying to start another container with the same name as an existing one.  Collection of the CIDs is only done to help with cleanup.

Reply:
Ahhhhh.... no, I got it now. Here, this should shed light:
```console
$ for i in 1 2 3;do bin/podman run -d quay.io/libpod/testimage:20221018 top >/dev/null; bin/podman inspect -l --format '{{.Name}}';done
gifted_bhabha
magical_goldwasser
gifted_thompson

! clean up
$ bin/podman rm -f -a >&/dev/null
$ !for
gifted_bhabha
magical_goldwasser
gifted_thompson
```
(that's just with my `1024` diff above. The `vendor` file is irrelevant).

So, there must be disambiguation code in podman name selection. That seems sensible. And we have confirmed that with a same seed, the **SEQUENCE** of names is what is reproducible. Finding a test for that is left as an exercise for the reader.

Reply:
> continue to try generating new names until there is no conflict

Oooohhh, so generating duplicate names to existing containers is impossible then.  In that case the new test is useless.

Reply:
okay, test is removed.

Reply:
>a test could create a container, record the name, delete it; create another container, record the name, delete it. Then check that the two generated names are different

Exactly.
```console
$ for i in 1 2 3;do bin/podman run quay.io/libpod/testimage:20221018 true; bin/podman inspect -l --format '{{.Name}}';bin/podman rm -l &>/dev/null;done
gifted_bhabha
gifted_bhabha
gifted_bhabha
```

Reply:
Okay ya, there you go.  That was with your fixed-seed though ya?   Hmmm, could be an worthwhile e2e test to have then?

Reply:
Yes, with fixed-seed. I don't know enough to comment on how useful a test would be.

Reply:
Well, IMHO under some use-case it seems it might be a fairly bad thing if podman started reusing names (or same sequence of names) like that.  The changes in this PR **DO** touch the RNG initialization code.  So perhaps it's worthwhile having a test.  Maybe let's do that as a followup PR though, this one is getting long in the tooth.

Reply:
Ideally, all code should be reliable and robust against conflicts. I think that’s the case for the specific target of `generateNames` (or at least that one loop lined above), but it’s historically not the case for _all_ the random names the codebase generates (I think something in the link cluster around https://github.com/containers/storage/pull/1337 contains examples) — sometimes, on a conflict, we just fail.

And even if the loops do exist, they are rarely-exercised code that might fail, compare a recent https://github.com/containers/buildah/issues/4679 .

And the conflicts are more costly in performance (e.g. container creation time, which we have focused on recently) than not having conflicts.

So, on net, I would say that there is _some_ value to us having a test that flags if we ever regress to repeated name sequences. I don’t have much of an opinion on how much effort to spend on it, or, more importantly, if we can only cause a potentially-flaky test that depends on a RNG seed (with some ~2^-32?? probability), whether that flake rate increase is worth it.

Reply:
Opened https://github.com/containers/podman/pull/18381